### DISCUSSION TITLE
운영체제 - 프로세스 동기화&데드락 - 호제&진호

### DISCUSSION THREAD

**Discussion by Jihn0118** [Shortudy-10th/tech-for-developer#141] [질의 응답]
### 프로세스 동기화 & 데드락 1
- 세마포어와 뮤텍스의 차이점은 무엇인가요?
- 데드락의 4가지 필요 조건은 무엇인가요?
- 프로세스 동기화가 무엇이고, 프로세스 동기화를 하는 이유에 대해 설명해주세요.
- 교착 상태가 무엇인가요?
- 공유 자원이란 무엇인가요?
---
### 프로세스 동기화 & 데드락 2
- 임계 구역이란 무엇인가요?
- 임계 구역 해결 조건에 대해 설명해주세요.
- 교착상태 해결 방법의 종류에 대해 말해주세요.
- 회피 기법인 은행원 알고리즘에 대해 설명해보세요.
- Race Condition(경쟁 상태)가 무엇이며, 왜 문제가 되는지 설명해주세요.

**Comment by Jihn0118** [comment]
## ❓️ 교착 상태가 무엇인가요?

💬 답변
교착상태(Dead Lock)는 상호 배제에 의해 나타나는 문제점으로, 
**둘 이상의 프로세스들이 자원을 점유한 상태에서 서로 다른 프로세스가 점유하고 있는 자원을 요구하며 무한정 기다리는 현상**입니다.

**Comment by Jihn0118** [comment]
## ❓️ DeadLock(교착상태)의 4가지 필요 조건은 무엇인가요?

💬 답변
### 1. 상호 배제(Mutual Exclusion)
- 한 번에 한 개의 프로세스만이 공유 자원을 사용할 수 있습니다.

### 2. 점유와 대기(Hold and Wait)
- 프로세스가 할당된 자원을 가진 상태에서 다른 자원을 기다립니다.

### 3. 비선점(Non-preemption)
- 프로세스가 작업을 마친 후 자원을 자발적으로 반환할 때까지 기다립니다.
- 이미 할당된 자원을 강제적으로 빼앗을 수 없습니다.

### 4. 환형 대기(Circular Wait)
- 프로세스의 자원 점유 및 점유 된 자원의 요구 관계가 원형을 이루면서 대기하는 조건입니다.
- 각 프로세스는 순환적으로 다음 프로세스가 요구하는 자원을 가지고 있습니다.

위 4가지 조건이 모두 만족 되는 경우 교착 상태 발생할 수 있습니다.
네 가지 조건 중 하나라도 충족되지 않으면 교착 상태가 발생하지 않습니다.

**Comment by Jihn0118** [comment]
## ❓️ 세마포어(Semaphore)와 뮤텍스(Mutex)의 차이점은 무엇인가요?

💬 답변

### 세마포어 vs 뮤텍스
뮤텍스는 한 번에 하나의 쓰레드만이 자원에 접근할 수 있도록 보장하며,
세마포어는 동시에 접근을 허용하는 쓰레드 수를 제어하여 병럴 처리를 조절합니다.

**Comment by zini9188** [comment]
❓️ 프로세스 동기화가 무엇이고, 프로세스 동기화를 하는 이유에 대해 설명해주세요.

💬 답변
여러 프로세스나 스레드가 공유 자원에 동시 접근하는 것을 제어하고, 서로의 작업을 조율하기 위한 기법이다. 주로 멀티 프로세스 혹은 멀티 스레드 환경에서 발생하는 문제를 해결하기 위해 사용된다.

프로세스 동기화를 하는 가장 큰 이유는 `임계 구역`문제 때문이다. 
여러 프로세스가 임계 구역에 동시 접근하면 데이터의 비일관성이 발생할 수 있고 데이터 훼손이 발생할 수 있다. 이를 경쟁 상태라고 하며 프로세스 동기화를 통해 데이터의 일관성을 유지할 수 있다.

**Reply by zini9188** [reply]
비선점

**Comment by zini9188** [comment]
❓️ 공유 자원이란 무엇인가요?

💬 답변
- 여러 프로세스 또는 스레드가 동시에 접근하거나 사용할 수 있는 데이터나 리소스
- 프로세스나 스레드가 정보를 주고받거나, 서로 협력하여 작업을 수행하기 위해 공유 자원을 사용
- 공유 자원의 종류
  - 변수, 메모리, 파일

- 공유 자원에 동시에 접근하면서 '경쟁 상태(Race Condition)'문제가 발생할 수 있음
  - '프로세스 동기화'를 통해 여러 프로세스나 스레드의 실행을 서로 조율
    - 공유 자원에 안전하게 접근 가능하고, 프로그램의 안정성 및 성능이 향상
